#Домашнє завдання до тем apply(), groupby(), pivot_table()

В цьому домашньому завданні продовжуємо працювати з набором даних `supermarket_sales.csv`.

0. Імпортуйте бібліотеку pandas та зчитайте дані у змінну `df` типу `pandas.DataFrame`.

In [1]:
import pandas as pd

from google.colab import drive
drive.mount('/content/drive')

df = pd.read_csv('drive/MyDrive/DataAnalytics_Course/Module_10_Pandas_Numpy/Data/supermarket_sales.csv')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


1. Дослідимо, який філіал супермаркету ('Branch') є найприбутковішим. Для цього знайдіть сумарний прибуток за кожним філіалом і виявіть, який філіал має найвищий.

In [2]:
df.columns #щоби бачити точні назви колонок

Index(['Invoice ID', 'Branch', 'City', 'Customer type', 'Gender',
       'Product line', 'Unit price', 'Quantity', 'Tax 5%', 'Total', 'Date',
       'Time', 'Payment', 'cogs', 'gross margin percentage', 'gross income',
       'Rating'],
      dtype='object')

In [3]:
df.groupby(['Branch'])['Total'].sum()
#найвищий прибуток має філіал С

,Total
Branch,
A,106200.3705
B,106197.6720
C,110568.7065


2. В якому місті знайходиться філіал з найвищим прибутком? Може в тому місці нам розмітисти ще один магазин.  
Знайдіть відповідь за допомогою функціоналу Pandas.

In [4]:
df[(df['Branch'] == "C")][['Branch', 'City']].head(1)


,Branch,City
1,C,Naypyitaw


In [5]:
df.groupby(["Branch"])['Customer type'].value_counts(normalize=True)
# у філіалі C найбільше купили товарів покупці, які мають членство (51.5%), тоді як в інших супермаркетах
# частіше купляли ті, щр членства не мають. Що говорить про більшу лояльність покупців у філіалі C
# цей факт може бути "за" відкриття нового магазину в місті Naypyitaw

Branch  Customer type
A       Normal           0.508824
        Member           0.491176
B       Normal           0.503012
        Member           0.496988
C       Member           0.515244
        Normal           0.484756
Name: proportion, dtype: float64

In [6]:
df.groupby(["Branch",'Customer type'])['Total'].sum().round(0)
# додаткова перевірка лояльності покупців, що мають членство
# результат перевірки: найбільше товарів у всіх філіях купують покупці, що мають членство

Branch  Customer type
A       Member           53637.0
        Normal           52563.0
B       Member           53705.0
        Normal           52493.0
C       Member           56881.0
        Normal           53687.0
Name: Total, dtype: float64

Як видно з 2х попередніх розрахунків, у філіалі С в місті Naypyitaw є найбільші продажі та найбільш лояльні покупці, бо 51.5% з них мають членство. Ці факти можуть говорити "за" відкриття ще одного магазину в місті Naypyitaw.

Але щоби дати точну рекомендацію, потрібно ще розрахувати необхідні капіталовкладення та операційні витрати, прогнозовані доходи та прибутки від 2го магазину.

Крім того, відкриття ще одного магазину несе додаткові операційні ризики як невизначеність якості менеджменту нового магазину, тоді як існуючий філіал є добре працюючою одиницею.

Відповідно необхідно перед прийняттям рішення порівняти опцію нового магазину з можливостями збільшенням продаж в існуючому філіалі.

3.1. Створіть зводну таблицю, яка покаже, скільки покупок (інвойсів) було зроблено в кожній з філій (`Branch`) за різними категоріями товарів. Запишіть таблицю в змінну `invoices_by_category` і виведіть змінну на екран.
Ця таблиця допоможе проаналізувати, в якій філії купують найбільше товарів кожної з категорій.

In [7]:
invoices_by_category = pd.pivot_table(df,index="Branch", columns='Product line', values='Invoice ID', aggfunc='count')
invoices_by_category

Product line,Electronic accessories,Fashion accessories,Food and beverages,Health and beauty,Home and lifestyle,Sports and travel
Branch,,,,,,
A,60,51,58,47,65,59
B,55,62,50,53,50,62
C,55,65,66,52,45,45


Очікуваний результат:

![](https://drive.google.com/uc?export=view&id=1rueAdko6S3UxIHGtojetTxlES-EyM6Yb)

3.2. Викристовуючи змінну `invoices_by_category` дайте відповідь програмно (тобто значення треба не просто знайти очима, а вивести за допомогою коду), в якому філіалі магазину (`Branch`) найбільше інвойсів із покупкою товарів категорії "Електронні аксесуари" (`Electronic accessories`)?


In [8]:
invoices_by_category['Electronic accessories'].sort_values(axis=0, ascending=False).head(1)

,Electronic accessories
Branch,
A,60


4-6. **Творче завдання на розвиток аналітичного мислення**

Крок 1. Сформулюйте ТРИ питання (гіпотези) до наявних даних, які допомогли б вам зрозуміти, які користувачі що, де та коли найбільше/найменше купують, аби дати на основі цих гіпотез рекомендації бізнесу. Звісно питання мають бути не тими, на які ми вже відповіли в завданнях модулю.

Крок 2. Знайдіть відповіді на свої питання з допомогою функціоналу pandas.

Крок 3. Напишіть, як відповідь на це питання може бути використана для прийняття бізнес рішень.   
   
 Питання можуть бути будь-якої складності, але їх має бути 3. Кожне питання оцінюється як 1 завдання. Без виконання цього завдання ДЗ не приймається. Якщо є питання щодо виконання - пишіть у чат 🙌


#Питання:

1) які категорії товарів приносять найбільший прибуток, а які найменший?

2) які категорії товарів найбільше купують жінки, а які чоловіки?

3) які методи оплати використовують найбільше?

In [47]:
# 1 питання: які категорії товарів приносять найбільший прибуток, а які найменший?

income_by_category = df.groupby(['Product line'])['gross income'].sum()
income_by_category.round()

,gross income
Product line,
Electronic accessories,2588.0
Fashion accessories,2586.0
Food and beverages,2674.0
Health and beauty,2343.0
Home and lifestyle,2565.0
Sports and travel,2625.0


Найбільший прибуток приносять Food and beverages, майже такий самий прибуток має категорія Sports and travel.

Найменший прибуток приносять Health and beauty.

Така інформація може бути корисною для прийняття рішень щодо категорій товарів, які потрібно й надалі підтримувати і розвивати, а на які звернути більше уваги та розробити маркетингову стратегію їх просування, або підкоригувати асортимент, або можливо й зняти з продажу.

В нашому випадку, наприклад, Health and beauty є кандидатом на більш ґрунтовний аналіз, щоби знайти можливу причину непопулярності серед покупців і вирішити як діяти, щоби збільшити її популярність.

Зняття з продажу є крайньою мірою, бо часто для супермаркетів важлива різноманітність товарів. і якщо наприклад прийняти рішення зняти з продажу категорію Health and beauty, то людина ймовірніше вибере інший супермаркет, щоби зробити свої регулярні покупки, де є ця категорія товарів в продажу.

In [44]:
# 2 питання: які категорії товарів найбільше купують жінки, а які чоловіки?
revenue_by_gender = pd.pivot_table(df,index="Product line", columns='Gender', values='Total', aggfunc='sum')
revenue_by_gender.loc['Grand Total'] = revenue_by_gender.sum(numeric_only=True).values
revenue_by_gender.round()

Gender,Female,Male
Product line,,
Electronic accessories,27102.0,27236.0
Fashion accessories,30437.0,23868.0
Food and beverages,33171.0,22974.0
Health and beauty,18561.0,30633.0
Home and lifestyle,30037.0,23825.0
Sports and travel,28575.0,26548.0
Grand Total,167883.0,155084.0


Жінки найбільше зробили покупок із категорії Food and beverages, а чоловіки - Health and beauty(що є неочікувано).

Наступні по популярності у жінок категорії: Fashion accessories та Home and lifestyle. Популярність Food and beverages, Fashion accessories та Home and lifestyle у жінок є цілком очікуваною.

У чоловіків також популярні категорії: Electronic accessories та Sports and travel.

Найменш популярні категорії у жінок - Health and beauty, а чоловіків - Food and beverages.

Головний інсайт - категорія Health and beauty, яка є найпопулярнішою у чоловіків і найменш популярною серед жінок.

Такв інформація може використовуватися для більш таргетованих пропозицій: різних для жінок та чоловіків, а також для аналізу та можливому перегляду асортименту товарів. Наприклад, щодо нашого інсайту треба зрозуміти чому чоловіки купують товари із Health and beauty більше за жінок. Тут може схований суттєвий додатковий ріст продажів.

In [37]:
# 3 питання: які методи оплати використовують найбільше?
df.groupby(['Payment'])['Total'].agg(['count','sum','mean'])


,count,sum,mean
Payment,,,
Cash,344,112206.570,326.181890
Credit card,311,100767.072,324.009878
Ewallet,345,109993.107,318.820600


Найчастіше покупці використовують метод оплати Ewallet (345 разів) та майже так само Cash (344 рази).
Але суми оплачених товарів готівкою вище, ніж Ewallet (112.2 тис иа 110 тис відповідно).
Найбільша середня оплата у метода Cash (326.18), а у метода Ewallet вона найменша (318.82)

Така інформація дає можливість бізнесу прийняти рішення щодо підтримання/ не підтримання методів оплат, а також щодо розвитку окремих методів оплат для збільшення виручки та прибутку.

В нашому випадку є доречним звернути увагу на метод Ewallet, який використовується так само часто, але має найменшу середню оплату. Але тут важливо обережно ставитись до висновків, щоби наприклад за додаткові бонуси за розрахунок Ewallet не сталося просте перетікання з одного методу оплати на інший, не створюючи додаткової вартості бізнесу.